concourse/client/util.py

```diff
@@ -122,3 +156,163 @@ def jobs_not_triggered(
+                pipeline_name=resource.pipeline_name(),
+                job_name=job.pipeline_name,
+            )
+
+            if wait_for_job_to_be_triggered(
+                job=job,
+                resource_version=resource_version,
+                concourse_api=concourse_api,
+                retries=retries_for_job_trigger,
+                sleep_time_seconds=sleep_seconds_between_attempts,
+            ):
+                info(f'{job.name=} for {resource_version.version()=} has been triggered')
+                return
+
+            else:
+                # job did not start, update pin comment with new retry timestamp
+                _pin_and_comment_resource(
+                    next_retry_in_seconds=retries_for_job_trigger * sleep_seconds_between_attempts,
+                    resource=resource,
+                    resource_version=resource_version,
+                    concourse_api=concourse_api,
+                )
+                warning(
+                    f'job {job.name} for resource version {resource_version.version()} '
+                    f'could not be triggered. Will retry...'
+                )
+
+        warning(
+            f'job {job.name} for resource version {resource_version.version()} '
+            f'could not be triggered. Giving up after {retries} retries'
+        )
+
+
+def _pin_and_comment_resource(
+    next_retry_in_seconds: int,
+    resource: PipelineConfigResource,
+    resource_version: ResourceVersion,
+    concourse_api,
+):
+    now = datetime.now()
+    next_retry = now + timedelta(seconds=next_retry_in_seconds)
+    pin_comment = PinComment(
+        pin_timestamp=now.isoformat(),
+        version=resource_version.version(),
+        next_retry=next_retry.isoformat(),
+        comment=(
+            "pinned by technical user 'concourse', please do not unpin before "
+            f"{next_retry.isoformat()}"
+        ),
+    )
+    concourse_api.pin_and_comment_resource_version(
+        pipeline_name=resource.pipeline_name(),
+        resource_name=resource.name,
+        resource_version_id=resource_version.id(),
+        comment=json.dumps(dataclasses.asdict(pin_comment)),
+    )
+
+
+@contextmanager
+def ensure_resource_unpinned(
+    resource: PipelineResource,
+    resource_version: ResourceVersion,
+    concourse_api,
+):
+    try:
+        _ensure_resource_unpinned(
+            resource=resource,
+            resource_version=resource_version,
+            concourse_api=concourse_api,
+        )
+        yield
+    finally:
+        concourse_api.unpin_resource(
+            pipeline_name=resource.pipeline_name(),
+            resource_name=resource.name,
+        )
+        info(f'successfully unpinned {resource.name=}')
+
+
+def _ensure_resource_unpinned(
+    resource: PipelineResource,
+    resource_version: ResourceVersion,
+    concourse_api,
+):
+    previous_retry_time = None
+    maximal_unpin_checktime = datetime.now() + timedelta(minutes=5)
```

Comment:
make this time a parameter (it is ok to add a default value, of course)?

also: maybe 5m are a bit too long..